src/watchdog.js

```diff
@@ -0,0 +1,414 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module watchdog/watchdog
+ */
+
+/* globals console, window */
+
+import mix from '@ckeditor/ckeditor5-utils/src/mix';
+import EmitterMixin from '@ckeditor/ckeditor5-utils/src/emittermixin';
+import { throttle, cloneDeepWith, isElement } from 'lodash-es';
+import CKEditorError from '@ckeditor/ckeditor5-utils/src/ckeditorerror';
+import areConnectedThroughProperties from '@ckeditor/ckeditor5-utils/src/areconnectedthroughproperties';
+
+/**
+ * A watchdog for CKEditor 5 editors.
+ *
+ * It keeps an {@link module:core/editor/editor~Editor editor} instance running.
+ * If a {@link module:utils/ckeditorerror~CKEditorError `CKEditorError` error}
+ * is thrown by the editor, it tries to restart the editor to the state before the crash. All other errors
+ * are transparent to the watchdog. By looking at the error context, the Watchdog restarts only the editor which crashed.
+ *
+ * Note that the watchdog does not handle errors during editor initialization (`Editor.create()`)
+ * and editor destruction (`editor.destroy()`). Errors at these stages mean that there is a serious
+ * problem in the code integrating the editor and such problem cannot be easily fixed restarting the editor.
+ *
+ * A new editor instance is created each time the watchdog is restarted. Thus the editor instance should not be kept
+ * internally. Use the `watchdog.editor` each time you need to access the editor.
+ *
+ * Basic usage:
+ *
+ * 		const watchdog = Watchdog.for( ClassicEditor );
+ *
+ * 		watchdog.create( elementOrData, editorConfig ).then( editor => {} );
+ *
+ * Full usage:
+ *
+ *		const watchdog = new Watchdog();
+ *
+ *		watchdog.setCreator( ( elementOrData, editorConfig ) => ClassicEditor.create( elementOrData, editorConfig ) );
+ *		watchdog.setDestructor( editor => editor.destroy() );
+ *
+ *		watchdog.create( elementOrData, editorConfig ).then( editor => {} );
+ *
+ * Other important APIs:
+ *
+ *		watchdog.on( 'error', () => { console.log( 'Editor crashed.' ) } );
+ *		watchdog.on( 'restart', () => { console.log( 'Editor was restarted.' ) } );
+ *
+ *		watchdog.restart(); // Restarts the editor.
+ *		watchdog.destroy(); // Destroys the editor and global error event listeners.
+ *
+ * 		watchdog.editor; // Current editor instance.
+ *
+ * 		watchdog.crashes.forEach( crashInfo => console.log( crashInfo ) );
+ */
+export default class Watchdog {
+	/**
+	 * @param {Object} config The watchdog plugin configuration.
+	 * @param {Number} config.crashNumberLimit A threshold specifying the number of crashes
+	 * when the watchdog stops restarting the editor in case of errors.
+	 * @param {Number} config.waitingTime A minimum amount of time between saving editor data internally.
+	 */
```

Comment:
The examples specified that you can create watchdog without any parameters. But `config` here is not marked as optional. I believe it is optional (a typical user will not know what kind of value is right here). If it is optional, please add default values in docs.

Reply:
I see that default values are specified in members description. Still, we need them here.